File: golang/lexer.go
package main

import (
    "fmt"
    "strings"
)

type Lexer struct {
  text string
  loc int
  cc byte // naive ascii
}


func NewLexer(text string) Lexer {
  lexer := Lexer{text: text, loc: 0}
  lexer.advance()
  return lexer
}

func (this Lexer) String() string {
  return fmt.Sprintf("{Lexer \"%s\" 0x%x @ %d}", this.text, this.cc, this.loc)
}

func (this *Lexer) NextToken() Token {
  for {
    for this.cc == ' ' || this.cc == '\n' || this.cc == '\t' || this.cc == '\r' {
      this.advance()
    }
    if this.cc != '#' {
      break
    }
    for this.cc != '\n' && this.cc != 0 {
      this.advance()
    }
  }
  if this.cc == 0 {
  return NewToken(EndOfFile, "")
  }
  if isalpha(this.cc) {
    return this.makeText() // private
  }
  if isdigit(this.cc) {
    return this.makeNumber()
  }
  if this.cc == '"' {
    return this.makeStringLiteral()
  }
  return this.makeSymbol()
}

func (this *Lexer) advance() {
  if this.loc < len(this.text) {
    this.cc = this.text[this.loc]
    this.loc++
  } else {
    this.cc = 0
  }
}

func (this *Lexer) makeSymbol() Token {
  first := this.cc
  value := string(first)
  this.advance()
  if this.cc == '=' && (first == '!' || first == '=' || first == '>' || first == '<') {
    value += string(this.cc)
    this.advance()
  }
  symbolType, ok := toSymbolType[value]
  if ok {
    return Token{tokenType: Symbol, value: value, symbol: symbolType}
  }
  panic("Cannot parse symbol " + value)
}


func (this *Lexer) makeText() Token {
  first := this.cc
  value := string(first)
  this.advance()
  if !isalpha(this.cc) {
    // next char is not alpha; it's a variable.
    varType := StrVarType
    switch {
      case (first >= 'a' && first <= 'h') || (first >= 'A' && first <= 'H'):
        varType = FloatVarType
      case (first >= 'i' && first <= 'n') || (first >= 'I' && first <= 'N'):
        varType = IntVarType
      default:
        varType = StrVarType
    }

    return Token{tokenType: Var, value: value, varType: varType}
  }

  for isalpha(this.cc) && this.cc != 0 {
    value += string(this.cc)
    this.advance()
  }
  value = strings.ToUpper(value)
  kw, ok := toKeywordType[value]
  if ok {
    return Token{tokenType: Keyword, value: value, keyword: kw}
  }
  panic("Unknown keyword " + value)
}

func (this *Lexer) makeNumber() Token {
  value := string(this.cc)
  this.advance()
  for isdigit(this.cc) && this.cc != 0 {
    value += string(this.cc)
    this.advance()
  }
  if this.cc == '.' {
    // Float
    value += string(this.cc)
    this.advance()
    for isdigit(this.cc) && this.cc != 0 {
      value += string(this.cc)
      this.advance()
    }
    return Token{tokenType: Const, value: value, varType: FloatVarType}
  }
  return Token{tokenType: Const, value: value, varType: IntVarType}
}

func (this *Lexer) makeStringLiteral() Token {
  this.advance() // eat the starting quote
  value := ""
  for this.cc != '"' && this.cc != 0 {
    value += string(this.cc)
    this.advance()
  }
  if this.cc == 0 {
    panic("Unclosed string literal")
  }
  this.advance() // eat the ending quote
  return Token{tokenType: Const, value: value, varType: StrVarType}
}

func isalpha(c byte) bool {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

func isdigit(c byte) bool {
  return c >= '0' && c <= '9'
}

File: c/lexer.c
#include <stdio.h>
#include <stdlib.h>
#include "lexer.h"
#include "sslc.h"
#include "token.h"


static Token *makeNumber(Lexer *this);
static Token *makeSymbol(Lexer *this);
static Token *makeText(Lexer *this);
static Token *makeString(Lexer *this);
static void advance(Lexer *this);

char *KEYWORDS[] = {
  "IF",
  "THEN",
  "ELSE",
  "ENDIF",
  "FOR",
  "TO",
  "STEP",
  "ENDFOR",
  "PRINT",
  "PRINTLN",
};


Lexer *newLexer(char *text) {
  Lexer *this = (Lexer *)calloc(sizeof(Lexer), 1);
  this->text = text;
  this->loc = text;
  advance(this);
  return this;
}


void advance(Lexer *this) {
  char *loc = this->loc;
  if (*loc != 0) {
    this->cc = *loc;
  } else {
    this->cc = 0;
  }
  this->loc++;
}


Token *nextToken(Lexer *this) {
  while (1) {
    while (this->cc <= ' ' && this->cc != 0) { // anything below 32 is whitespace.
      advance(this);
    }
    if (this->cc != '#') {
      break;
    }
    while (this->cc != '\n' && this->cc != 0) {
      advance(this);
    }
  }
  if (this->cc == 0) {
    return newToken(END_OF_FILE, NULL);
  }
  if (isdigit(this->cc)) {
    return makeNumber(this);
  }
  if (isalpha(this->cc)) {
    return makeText(this);
  }
  if (this->cc == '"') {
    return makeString(this);
  }
  return makeSymbol(this);
}


Token *makeString(Lexer *this) {
  char *start = this->loc;
  advance(this); // eat the starting quote
  while (this->cc != '"' && this->cc != 0) {
    advance(this);
  }
  if (this->cc == 0) {
    fail("Unclosed string literal");
  }
  advance(this); // eat the closing quote
  char *end = this->loc;
  int len = end - start - 1;
  char *literal = (char*)calloc(len, 0);
  strncpy(literal, start, len - 1);
  Token *t = newToken(CONST, literal);
  t->tokenData.varType = STR;
  return t;
}


Token *makeNumber(Lexer *this) {
  char *start = this->loc;
  while (isdigit(this->cc)) {
    advance(this);
  }
  VarType varType = INT;
  if (this->cc == '.') {
    advance(this);
    while (isdigit(this->cc)) {
      advance(this);
    }
    varType = FLOAT;
  }

  // now, *loc points at a non-digit
  int length = this->loc - start;
  char *value = (char*) calloc(length, 1);
  // subtract one because we started one past "first"
  strncpy(value, start - 1, length);
  Token *token = newToken(CONST, value);
  token->tokenData.varType = varType;
  return token;
}

Token *makeText(Lexer *this) {
  char *start = this->loc;
  char first = this->cc;
  advance(this);
  if (!isalpha(this->cc)) {
    // one letter 
    VarType varType = STR;
    if (first >= 'a' && first <= 'h') {
      varType = FLOAT;
    }
    if (first >= 'i' && first <= 'n') {
      varType = INT;
    }
    char *value = (char*) calloc(2, 1);
    value[0] = first;
    Token *token = newToken(VAR, value);
    token->tokenData.varType = varType;
    return token;
  }
  while (isalpha(this->cc)) {
    advance(this);
  }

  // this->loc points to the first non-alpha
  int length = this->loc - start;
  char *value = (char*) calloc(length, 1);
  // subtract one because we started one past "first"
  strncpy(value, start - 1, length);
  for (int i = 0; i < N_KEYWORDS; ++i) {
    if (stricmp(KEYWORDS[i], value) == 0) {
      Token *token = newToken(KEYWORD, value);
      token->tokenData.keyword = i;
      return token;
    }
  }
  fail("ERROR: Unknown keyword: %s", value);
  return NULL;
}

Token *makeSymbol(Lexer *this) {
  char first = this->cc;
  advance(this);
  if (first == '!') {
    if (this->cc == '=') {
      advance(this);
      Token *t = newToken(SYMBOL, "!=");
      t->tokenData.symbol = NEQ;
      return t;
    }
    fail("Unknown symbol !");
    return NULL;
  }

  SymbolType st = -1;
  char *value = NULL;
  // second char
  switch(first) {
    case '*':
      st = MULT;
      break;
    case '+':
      st = PLUS;
      break;
    case '-':
      st = MINUS;
      break;
    case '/':
      st = DIV;
      break;
    default:
      if (this->cc == '=') {
        switch (first) {
          case '=':
            st = EQEQ;
            value = "==";
            advance(this);
            break;
          case '>':
            st = GEQ;
            value =">=";
            advance(this);
            break;
          case '<':
            st = LEQ;
            value = "<=";
            advance(this);
            break;

          default:
            fail("Unknown symbol %c (%d) before =", first, first);
            break;
        }
      } else {
        switch (first) {
          case '=':
            st = EQ;
            break;
          case '>':
            st = GT;
            break;
          case '<':
            st = LT;
            break;
          default:
            fail("Unknown symbol %c (%d) not before =", first, first);
            break;
        }
      }
      break;
  }

  Token *t = newToken(SYMBOL, value);
  t->tokenData.symbol = st;
  return t;
}
